Package config with recursive option on folder with only sub-folders: index out of range error

PLEASE READ
-------------

DO NOT submit tickets without _first_ using the latest version of Golang, clearing your local golang package cache, and re-building mockery using the _latest_ Golang version and the _latest_ version of mockery. Please provide evidence this has been done in your issue. Failure to provide this evidence will likely result in your issue being closed.

Description
-------------

I have this configuration:
```
output: ./mocks/pkg
keeptree: true
case: underscore
packages:
  github.com/a-pedini/my-service/pkg: // this is just a dummy name as the repo is private
    config:
      all: true
      recursive: true
      with-expecter: true
```
and I don't have any files directly in the /pkg folders, just subfolders with sub-packages.

When I try to run it with this config I get:
```
docker run --rm -v "/Users/a-pedini/IdeaProjects/my-service":/src -v "/Users/a-pedini/.gvm/pkgsets/go1.20.3/global:/go" -v ":/tmp/.cache" -w /src vektra/mockery:v2.25.0 --config ./.mockery_test.yaml --log-level debug --dry-run
24 Apr 23 23:40 UTC INF Starting mockery dry-run=true version=v2.25.0
24 Apr 23 23:40 UTC INF Using config: ./.mockery_test.yaml dry-run=true version=v2.25.0
24 Apr 23 23:40 UTC DBG package cache is nil dry-run=true version=v2.25.0
24 Apr 23 23:40 UTC DBG discovering sub-packages dry-run=true package-name=github.com/a-pedini/my-service/pkg version=v2.25.0
panic: runtime error: index out of range [0] with length 0

goroutine 1 [running]:
github.com/vektra/mockery/v2/pkg/config.(*Config).subPackages(0xc0000a0a80?, {0xa502e0?, 0xc0001493b0?}, {0xc0000eca40, 0x20}, 0xc0000ef600?)
        /home/runner/work/mockery/mockery/pkg/config/config.go:441 +0x826
```
In order to make it work I have to create an empty go file in the /pkg level... I'm guessing that's because without this file, the folder pkg is not a real package? as there was no file with the package pkg header...

would it be possible to consider a folder, that I guess is not technically a package, as the starting point for the recursion? with the dummy file this works well and it is way faster than before, so I think we can also live with that but just checking :P

Mockery Version
--------

v2.25.0 via docker

Golang Version
--------------

1.20.3

```
NOTE: Please upgrade to the latest golang version before submitting tickets!
```

Installation Method
-------------------

- [ ] Binary Distribution
- [x] Docker
- [ ] brew
- [ ] go install
- [ ] Other: [specify]

Steps to Reproduce
------------------

1. Create a ./pkg folder
2. Add a subfolder inside ./pkg and add an interface to mock inside this subfolder
3. run mockery to create the mock

### Expected Behavior

Creates the mock in the /mocks/pkg/subfolder directory

### Actual Behavior

Index out of bounds error

--- a/Taskfile.yml
+++ b/Taskfile.yml
@@ -41,7 +41,7 @@ tasks:
       - go run github.com/golangci/golangci-lint/cmd/golangci-lint run
   
   test.ci:
-    deps: [test, fmt, mocks, lint]
+    deps: [fmt, lint, mocks, test]
 
   default:
     deps: [test.ci]

--- a/cmd/mockery.go
+++ b/cmd/mockery.go
@@ -36,13 +36,16 @@ func NewRootCmd() *cobra.Command {
 	cmd := &cobra.Command{
 		Use:   "mockery",
 		Short: "Generate mock objects for your Golang interfaces",
-		RunE: func(cmd *cobra.Command, args []string) error {
+		Run: func(cmd *cobra.Command, args []string) {
 			r, err := GetRootAppFromViper(viperCfg)
 			if err != nil {
 				printStackTrace(err)
-				return err
+				os.Exit(1)
 			}
-			return r.Run()
+			if err := r.Run(); err != nil {
+				printStackTrace(err)
+				os.Exit(1)
+			}
 		},
 	}
 
@@ -104,7 +107,6 @@ func printStackTrace(e error) {
 // Execute executes the cobra CLI workflow
 func Execute() {
 	if err := NewRootCmd().Execute(); err != nil {
-		// printStackTrace(err)
 		os.Exit(1)
 	}
 }

--- a/go.work.sum
+++ b/go.work.sum
@@ -510,6 +510,7 @@ github.com/shurcooL/go v0.0.0-20180423040247-9e1955d9fb6e h1:MZM7FHLqUHYI0Y/mQAt
 github.com/shurcooL/go-goon v0.0.0-20170922171312-37c2f522c041 h1:llrF3Fs4018ePo4+G/HV/uQUqEI1HMDjCeOf2V6puPc=
 github.com/spaolacci/murmur3 v0.0.0-20180118202830-f09979ecbc72 h1:qLC7fQah7D6K1B0ujays3HV9gkFtllcxhzImRR7ArPQ=
 github.com/spaolacci/murmur3 v0.0.0-20180118202830-f09979ecbc72/go.mod h1:JwIasOWyU6f++ZhiEuf87xNszmSA2myDM2Kzu9HwQUA=
+github.com/stretchr/testify v1.6.1/go.mod h1:6Fq8oRcR53rry900zMqJjRRixrwX3KX962/h/Wwjteg=
 github.com/stretchr/testify v1.7.5/go.mod h1:yNjHg4UonilssWZ8iaSj1OCr/vHnekPRkoO+kdMU+MU=
 github.com/tklauser/go-sysconf v0.3.11 h1:89WgdJhk5SNwJfu+GKyYveZ4IaJ7xAkecBo+KdJV0CM=
 github.com/tklauser/numcpus v0.6.0 h1:kebhY2Qt+3U6RNK7UqpYNA+tJ23IBEGKkB7JQBfDYms=

--- a/pkg/config/config.go
+++ b/pkg/config/config.go
@@ -20,8 +20,9 @@ import (
 )
 
 var (
-	ErrNoConfigFile = fmt.Errorf("no config file exists")
-	ErrPkgNotFound  = fmt.Errorf("package not found in config")
+	ErrNoConfigFile         = fmt.Errorf("no config file exists")
+	ErrNoGoFilesFoundInRoot = fmt.Errorf("no go files found in root search path")
+	ErrPkgNotFound          = fmt.Errorf("package not found in config")
 )
 
 type Interface struct {
@@ -427,7 +428,12 @@ func (c *Config) addSubPkgConfig(ctx context.Context, subPkgPath string, parentP
 	return nil
 }
 
-func (c *Config) subPackages(ctx context.Context, pkgPath string, pkgConfig *Config) ([]string, error) {
+func (c *Config) subPackages(
+	ctx context.Context,
+	pkgPath string,
+	pkgConfig *Config,
+	currentDepth int,
+) ([]string, error) {
 	log := zerolog.Ctx(ctx)
 
 	pkgs, err := packages.Load(&packages.Config{
@@ -438,6 +444,13 @@ func (c *Config) subPackages(ctx context.Context, pkgPath string, pkgConfig *Con
 	}
 	pkg := pkgs[0]
 
+	if currentDepth == 0 && len(pkg.GoFiles) == 0 {
+		log.Error().
+			Err(ErrNoGoFilesFoundInRoot).
+			Str("documentation", "https://vektra.github.io/mockery/notes/#error-no-go-files-found-in-root-search-path").
+			Msg("package contains no go files")
+		return nil, ErrNoGoFilesFoundInRoot
+	}
 	representativeFile := pathlib.NewPath(pkg.GoFiles[0])
 	searchRoot := representativeFile.Parent()
 	packageRootName := pathlib.NewPath(pkg.PkgPath)
@@ -498,7 +511,7 @@ func (c *Config) subPackages(ctx context.Context, pkgPath string, pkgConfig *Con
 		return nil
 	})
 	if walkErr != nil {
-		return nil, fmt.Errorf("error occured during filesystem walk: %w", err)
+		return nil, fmt.Errorf("error occured during filesystem walk: %w", walkErr)
 	}
 
 	// Parse the subdirectories we found into their respective fully qualified
@@ -538,10 +551,10 @@ func (c *Config) discoverRecursivePackages(ctx context.Context) error {
 		return nil
 	}
 	for pkgPath, conf := range recursivePackages {
-		pkgLog := log.With().Str("package-name", pkgPath).Logger()
+		pkgLog := log.With().Str("package-path", pkgPath).Logger()
 		pkgCtx := pkgLog.WithContext(ctx)
 		pkgLog.Debug().Msg("discovering sub-packages")
-		subPkgs, err := c.subPackages(pkgCtx, pkgPath, conf)
+		subPkgs, err := c.subPackages(pkgCtx, pkgPath, conf, 0)
 		if err != nil {
 			return fmt.Errorf("failed to get subpackages: %w", err)
 		}

--- a/pkg/config/config_test.go
+++ b/pkg/config/config_test.go
@@ -3,6 +3,7 @@ package config
 import (
 	"bytes"
 	"context"
+	"errors"
 	"reflect"
 	"testing"
 
@@ -787,8 +788,18 @@ func TestConfig_Initialize(t *testing.T) {
 		name       string
 		cfgYaml    string
 		wantCfgMap string
-		wantErr    bool
+		wantErr    error
 	}{
+		{
+			name: "package with no go files",
+			cfgYaml: `
+packages:
+  github.com/vektra/mockery/v2/pkg/fixtures/pkg_with_no_files:
+    config:
+      recursive: True
+      all: True`,
+			wantErr: ErrNoGoFilesFoundInRoot,
+		},
 		{
 			name: "test with no subpackages present",
 			cfgYaml: `
@@ -951,14 +962,14 @@ with-expecter: false
 			log, err := logging.GetLogger("TRACE")
 			require.NoError(t, err)
 
-			if err := c.Initialize(log.WithContext(context.Background())); (err != nil) != tt.wantErr {
+			if err := c.Initialize(log.WithContext(context.Background())); !errors.Is(err, tt.wantErr) {
 				t.Errorf("Config.Initialize() error = %v, wantErr %v", err, tt.wantErr)
 			}
 
 			cfgAsStr, err := yaml.Marshal(c._cfgAsMap)
 			require.NoError(t, err)
 
-			if !reflect.DeepEqual(string(cfgAsStr), tt.wantCfgMap) {
+			if tt.wantCfgMap != "" && !reflect.DeepEqual(string(cfgAsStr), tt.wantCfgMap) {
 				t.Errorf(`Config.Initialize resultant config map
 got
 ----

--- a/pkg/fixtures/pkg_with_no_files/subpkg/foo.go
+++ b/pkg/fixtures/pkg_with_no_files/subpkg/foo.go
@@ -1,0 +1,1 @@
+package foo

